### PR TITLE
Hook up mgo logging and reconnect properly.

### PR DESCRIPTION
**Pull request by howbazaar**
Now the mgo logging is output through loggo to the "mgo" logger.
This means it is now available to be shown, but not shown by default.
To enable in other tests, they can (after logging is set up) set:
  loggo.GetLogger("mgo").SetLogLevel(loggo.TRACE)
Also, when the mongo server is changed by either restarting, or other tests killing it, the teardown now reconnects the session, which avoids some of the "no reachable servers" errors we were seeing.

### CODE REVIEW

File: mgo.go [hunk 1]
```diff
@@ -422,7 +423,16 @@ func MgoTestPackage(t *testing.T, certs *Certs) {
 	gc.TestingT(t)
 }
 
+// Output implements the mgo log_Logger interface.
+func (s *MgoSuite) Output(calldepth int, message string) error {
```
Comment (mjs) [inline mgo.go]: I don't mind too much, but this doesn't really belong on the suite. A simpler wrapper type would be cleaner. The logger wouldn't even have to be added as a field on MgoSuite...

File: mgo.go [hunk 2]
```diff
@@ -772,7 +785,16 @@ func (s *MgoSuite) TearDownTest(c *gc.C) {
 	err := MgoServer.EnsureRunning()
 	c.Assert(err, jc.ErrorIsNil)
 
-	if err = s.Session.Ping(); err != nil {
+	// If the Session we have doesn't know about
+	// the address of the server, then we should reconnect.
+	foundAddress := false
+	for _, addr := range s.Session.LiveServers() {
+		if addr == MgoServer.Addr() {
+			foundAddress = true
```
Comment (mjs) [inline mgo.go]: break here?

File: mgo.go [hunk 3]
```diff
@@ -758,6 +770,7 @@ func (inst *MgoInstance) EnsureRunning() error {
 	// If the server has already been destroyed for testing purposes,
 	// just start it again.
 	if inst.Addr() == "" {
+		logger.Infof("restarting mongo instance")
```
Comment (mjs) [inline mgo.go]: this should probably be debug

File: mgo.go (hunk 1 above)
Comment (howbazaar) [inline mgo.go]: Fixed.

File: mgo.go (hunk 3 above)
Comment (howbazaar) [inline mgo.go]: Fixed.

File: mgo.go (hunk 2 above)
Comment (howbazaar) [inline mgo.go]: Added.

**Comment by howbazaar** [comment]
$$merge$$

**Pull request merged**